### ISSUE TITLE
Update records in MySQL database not working

### ISSUE THREAD

**Issue by chrisis2015** [fsprojects/SQLProvider#384]
### Description

I try to update a record in a mysql database table but I get an error message:

The given key was not present in the dictionary

Reading and Inserting records is no problem. 

### Repro steps

I have this mysql table:

CREATE TABLE `SENSORLOCATION` (
  `ID` int(11) NOT NULL AUTO_INCREMENT,
  `LOCATION_ID` int(11) NOT NULL,
  `BTCODE` varchar(20) NOT NULL,
  `LOCCODE` varchar(20) NOT NULL,
  `LGROUP` int(11) NOT NULL,
  `LOCATIONNAME` varchar(50) NOT NULL,
  `LOCATIONPLACE` varchar(50) NOT NULL,
  `ACTIVE_EXPORT` varchar(1) NOT NULL,
  `TYPECONN` int(11) NOT NULL COMMENT '0= VPN, 1= HTTP, 2 = EMIL',
  `ISENTRANCE` tinyint(1) NOT NULL,
  `LAST_DATE_TIME` datetime NOT NULL,
  `REGIO` int(11) NOT NULL DEFAULT '1',
  `ACTIVE` tinyint(1) NOT NULL DEFAULT '1',
  `CONVERSION` double NOT NULL DEFAULT '1',
  `HASHED_VERSION` varchar(1) NOT NULL DEFAULT '0',
  `EMAIL` varchar(100) DEFAULT NULL,
  `TEL` varchar(45) DEFAULT NULL,
  `SENSORNAME` varchar(45) NOT NULL DEFAULT 'Custom Scanner',
  `REMARK` varchar(200) DEFAULT ' ',
  `REMARK_DATE` datetime DEFAULT NULL,
  `NOTIFIED_OFFLINE` tinyint(1) NOT NULL DEFAULT '0',
  PRIMARY KEY (`ID`),
  UNIQUE KEY `ID_UNIQUE` (`ID`),
  KEY `FK_LGROUP_SL_SLG_idx` (`LGROUP`),
  KEY `IDX_BTCODE` (`BTCODE`,`LOCATION_ID`,`ACTIVE`)
) ENGINE=InnoDB AUTO_INCREMENT=491 DEFAULT CHARSET=latin1;

I use this code for updating table data:

    type dbSchema = SqlDataProvider<
                        ConnectionString = "validConnectionString",
                        DatabaseVendor = Common.DatabaseProviderTypes.MYSQL, 
                        ResolutionPath = @"valid resolution path", 
                        IndividualsAmount = 1000, 
                        Owner = "valid owner",
                        UseOptionTypes = true>

    let db = dbSchema.GetDataContext()

    let updateRecords = 
        query {
            for s in db.dbSensors.Sensorlocation do
            select s
        } 
        |> Seq.toList
        |> List.iter (fun s -> s.NotifiedOffline <- (sbyte) 1)       
        db.SubmitUpdates()

### Expected behavior

I would expect that column NotifiedOffline is updated to value 1 for all records in the table.

### Actual behavior

I get an error message:

The given key was not present in the dictionary

### Known workarounds

I have no workaround yet.

### Related information 

* Used database: MySQL
* Operating system: Windows 8

**Comment by Thorium** [comment]
Is s.SetColumn("NotifiedOffline", 1 |> box) working?

**Comment by chrisis2015** [comment]
Hi Thorium,

I tried it, it says:

error FS0001: Type mismatch. Expecting a
    int -> sbyte    
but given a
    int -> obj    
The type 'sbyte' does not match the type 'obj'

**Comment by Thorium** [comment]
Ok, I think this is a problem with tinyint to .Net type mapping. Can you update Regio or Lgroup columns?

**Comment by chrisis2015** [comment]
This results in the same error:

```
let updateRecords = 
    query {
        for s in db.dbSensors.Sensorlocation do
        select s
    } 
    |> Seq.toList
    |> List.iter (fun s -> s.SetColumn("Regio", 1 |> box))       
    db.SubmitUpdates()
```

Also tried:

```
    let updateRecords = 
        query {
            for s in db.BtadminBtScannerTool.Sensorlocation do
            select s
        } 
        |> Seq.toList
        |> List.iter (fun s -> s.SetColumn("Regio", uint32 1))       
        db.SubmitUpdates()
```

Again the error is the same, now added it with more tracing:

```
System.Collections.Generic.KeyNotFoundException: The given key was not present in the dictionary.
   at System.Collections.Generic.Dictionary`2.get_Item(TKey key)
   at FSharp.Data.Sql.Providers.MySqlProvider.createUpdateCommand(IDbConnection con, StringBuilder sb, SqlEntity entity, FSharpList`1 changedColumns)
   at <StartupCode$FSharp-Data-SqlProvider>.$Providers.MySql.FSharp-Data-Sql-Common-ISqlProvider-ProcessUpdates@634-9.Invoke(SqlEntity e)
   at Microsoft.FSharp.Collections.SeqModule.Iterate[T](FSharpFunc`2 action, IEnumerable`1 source)
   at FSharp.Data.Sql.Providers.MySqlProvider.FSharp-Data-Sql-Common-ISqlProvider-ProcessUpdates(IDbConnection con, ConcurrentDictionary`2 entities)
   at FSharp.Data.Sql.Runtime.SqlDataContext.FSharp-Data-Sql-Common-ISqlDataContext-SubmitPendingChanges()
```

**Comment by Thorium** [comment]
I just tried this with MySQL and it's working perfectly.
What version of SQLProvider you are using?

**Comment by chrisis2015** [comment]
I use version 1.0.6.0:

![image](https://cloud.githubusercontent.com/assets/26362002/23846626/9f2666ba-07ce-11e7-94d0-f34560614640.png)

And the version of mysql server used is not the newest either, it is version 5.1.73. I'm forced to use this older version for now.

**Comment by Thorium** [comment]
This is a bug in SQLProvider 1.0.6 which is fixed since then.
MySQL-version shouldn't matter.

**Comment by chrisis2015** [comment]
Great, I will try it out asap. Thanks for the help! I'll post the results later this week.

**Comment by chrisis2015** [comment]
I updated to the latest version (1.0.45.0) and tried again to update a record. Now the error is different:

```
System.Exception: Error - you cannot update an entity that does not have a primary key. (btadmin_bt_scannerTool.SENSORLOCATION)
   at FSharp.Data.Sql.Providers.MySqlProvider.createUpdateCommand(IDbConnection con, StringBuilder sb, SqlEntity entity, FSharpList`1 changedColumns)
   at <StartupCode$FSharp-Data-SqlProvider>.$Providers.MySql.FSharp-Data-Sql-Common-ISqlProvider-ProcessUpdates@722-3.Invoke(SqlEntity e)
   at Microsoft.FSharp.Collections.SeqModule.Iterate[T](FSharpFunc`2 action, IEnumerable`1 source)
   at FSharp.Data.Sql.Providers.MySqlProvider.FSharp-Data-Sql-Common-ISqlProvider-ProcessUpdates(IDbConnection con, ConcurrentDictionary`2 entities, TransactionOptions transactionOptions)
   at <StartupCode$FSharp-Data-SqlProvider>.$SqlRuntime.DataContext.f@1-32(SqlDataContext __, IDbConnection con, Unit unitVar0)
   at FSharp.Data.Sql.Runtime.SqlDataContext.FSharp-Data-Sql-Common-ISqlDataContext-SubmitPendingChanges()
```
The strange this is, as you can see in the create script for the table, there is a primary key column 'ID'.

**Comment by Thorium** [comment]
Ok, sounds like MySQL has changed it's way to handle primary keys, and we support only the new style. If we don't know the primary key, we don't know which entity to update, so we need the keys. Currently we are executing this SQL to get the keys:

```sql
SELECT DISTINCTROW c.COLUMN_NAME,c.DATA_TYPE, c.character_maximum_length, c.numeric_precision, c.is_nullable
            ,CASE WHEN pk.COLUMN_NAME IS NOT NULL THEN 'PRIMARY KEY' ELSE '' END AS KeyType
FROM INFORMATION_SCHEMA.COLUMNS c
LEFT JOIN (
            SELECT ku.TABLE_CATALOG,ku.TABLE_SCHEMA,ku.TABLE_NAME,ku.COLUMN_NAME
            FROM INFORMATION_SCHEMA.TABLE_CONSTRAINTS AS tc
            INNER JOIN INFORMATION_SCHEMA.KEY_COLUMN_USAGE AS ku
                ON tc.CONSTRAINT_TYPE = 'PRIMARY KEY'
                AND tc.CONSTRAINT_NAME = ku.CONSTRAINT_NAME
        )   pk
ON  c.TABLE_CATALOG = pk.TABLE_CATALOG
            AND c.TABLE_SCHEMA = pk.TABLE_SCHEMA
            AND c.TABLE_NAME = pk.TABLE_NAME
            AND c.COLUMN_NAME = pk.COLUMN_NAME
WHERE c.TABLE_SCHEMA = @schema AND c.TABLE_NAME = @table
```

Does that return anything, e.g. if you try to execute the SQL without the last where-condition?

**Comment by chrisis2015** [comment]
When I execute the query above like this:

```
SELECT DISTINCTROW c.COLUMN_NAME,c.DATA_TYPE, c.character_maximum_length, c.numeric_precision, c.is_nullable
            ,CASE WHEN pk.COLUMN_NAME IS NOT NULL THEN 'PRIMARY KEY' ELSE '' END AS KeyType
FROM INFORMATION_SCHEMA.COLUMNS c
LEFT JOIN (
            SELECT ku.TABLE_CATALOG,ku.TABLE_SCHEMA,ku.TABLE_NAME,ku.COLUMN_NAME
            FROM INFORMATION_SCHEMA.TABLE_CONSTRAINTS AS tc
            INNER JOIN INFORMATION_SCHEMA.KEY_COLUMN_USAGE AS ku
                ON tc.CONSTRAINT_TYPE = 'PRIMARY KEY'
                AND tc.CONSTRAINT_NAME = ku.CONSTRAINT_NAME
        )   pk
ON  c.TABLE_CATALOG = pk.TABLE_CATALOG
            AND c.TABLE_SCHEMA = pk.TABLE_SCHEMA
            AND c.TABLE_NAME = pk.TABLE_NAME
            AND c.COLUMN_NAME = pk.COLUMN_NAME
WHERE c.TABLE_SCHEMA = 'dbSensors' AND c.TABLE_NAME = 'SENSORLOCATION';
```

I got this result set:

![image](https://cloud.githubusercontent.com/assets/26362002/23871322/28a76348-082a-11e7-9c2f-6ad14c9ad23d.png)

Or is it only useful when I remove the where condition? Because above I referred to the specific table.

**Comment by Thorium** [comment]
Why your KeyType column is empty for ID?

After I ran your script, my query shows:
```
+------------------+-----------+--------------------------+-------------------+-------------+-------------+
| COLUMN_NAME      | DATA_TYPE | character_maximum_length | numeric_precision | is_nullable | KeyType     |
+------------------+-----------+--------------------------+-------------------+-------------+-------------+
| ID               | int       |                     NULL |                10 | NO          | PRIMARY KEY |
| LOCATION_ID      | int       |                     NULL |                10 | NO          |             |
| BTCODE           | varchar   |                       20 |              NULL | NO          |             |
| LOCCODE          | varchar   |                       20 |              NULL | NO          |             |
| LGROUP           | int       |                     NULL |                10 | NO          |             |
| LOCATIONNAME     | varchar   |                       50 |              NULL | NO          |             |
| LOCATIONPLACE    | varchar   |                       50 |              NULL | NO          |             |
| ACTIVE_EXPORT    | varchar   |                        1 |              NULL | NO          |             |
| TYPECONN         | int       |                     NULL |                10 | NO          |             |
| ISENTRANCE       | tinyint   |                     NULL |                 3 | NO          |             |
| LAST_DATE_TIME   | datetime  |                     NULL |              NULL | NO          |             |
| REGIO            | int       |                     NULL |                10 | NO          |             |
| ACTIVE           | tinyint   |                     NULL |                 3 | NO          |             |
| CONVERSION       | double    |                     NULL |                22 | NO          |             |
| HASHED_VERSION   | varchar   |                        1 |              NULL | NO          |             |
| EMAIL            | varchar   |                      100 |              NULL | YES         |             |
| TEL              | varchar   |                       45 |              NULL | YES         |             |
| SENSORNAME       | varchar   |                       45 |              NULL | NO          |             |
| REMARK           | varchar   |                      200 |              NULL | YES         |             |
| REMARK_DATE      | datetime  |                     NULL |              NULL | YES         |             |
| NOTIFIED_OFFLINE | tinyint   |                     NULL |                 3 | NO          |             |
+------------------+-----------+--------------------------+-------------------+-------------+-------------+
```

**Comment by chrisis2015** [comment]
This is strange, I'll try to find out the cause for the difference you and I got.

**Comment by Thorium** [comment]
Assuming that you have the primary key in place, maybe old MySQL 5.1 is managing the key column info somewhere else than `INFORMATION_SCHEMA.KEY_COLUMN_USAGE` table, the [documentation](https://dev.mysql.com/doc/refman/5.5/en/key-column-usage-table.html) is from 5.5. 

If you manage to find a way to identify the key column, we could modify this SQL-clause to get the keys also for older versions and then the update would start working.

**Comment by chrisis2015** [comment]
I've looked at your query for table info and tried to find a cause for not showing the primary key in the KeyType column in my case. 

The reason is that the columns INFORMATION_SCHEMA.KEY_COLUMN_USAGE.TABLE_CATALOG and INFORMATION_SCHEMA.COLUMNS.TABLE_CATALOG are both null and those fields are used in the last join-condition.

So when I remove that part of the join-condition, like this:

```
SELECT DISTINCTROW c.COLUMN_NAME,c.DATA_TYPE, c.character_maximum_length, c.numeric_precision, c.is_nullable
            ,CASE WHEN pk.COLUMN_NAME IS NOT NULL THEN 'PRIMARY KEY' ELSE '' END AS KeyType
FROM INFORMATION_SCHEMA.COLUMNS c
LEFT JOIN (
            SELECT ku.TABLE_CATALOG,ku.TABLE_SCHEMA,ku.TABLE_NAME,ku.COLUMN_NAME
            FROM INFORMATION_SCHEMA.TABLE_CONSTRAINTS AS tc
            INNER JOIN INFORMATION_SCHEMA.KEY_COLUMN_USAGE AS ku
                ON tc.CONSTRAINT_TYPE = 'PRIMARY KEY'
                AND tc.CONSTRAINT_NAME = ku.CONSTRAINT_NAME
        )   pk
ON  /*c.TABLE_CATALOG = pk.TABLE_CATALOG
            AND*/ c.TABLE_SCHEMA = pk.TABLE_SCHEMA
            AND c.TABLE_NAME = pk.TABLE_NAME
            AND c.COLUMN_NAME = pk.COLUMN_NAME
WHERE c.TABLE_SCHEMA = 'dbSensors' AND c.TABLE_NAME = 'SENSORLOCATION';
```
The result set is:

![image](https://cloud.githubusercontent.com/assets/26362002/23957356/73d83a10-099f-11e7-85ae-620568fb94f2.png)

Why the table_catalog column values are null in my case is not clear yet, but it solves part of the mystery I guess.

**Comment by Thorium** [comment]
Hmm... Will you still get all the keys for this: 
```sql
select * from INFORMATION_SCHEMA.KEY_COLUMN_USAGE;
```
...and the `TABLE_CATALOG` is null?

I hope many won't use table catalogs then if we modify the code... Maybe we could say:
```sql
... ON  (c.TABLE_CATALOG = pk.TABLE_CATALOG OR pk.TABLE_CATALOG IS NULL)
            AND c.TABLE_SCHEMA = pk.TABLE_SCHEMA ...
```

**Comment by chrisis2015** [comment]
When I execute that query, yes I get all the keys and TABLE_CATALOG is null in every record. 

Your solution could be an option, but I still have to find out exactly what is a catalog in the MySQL context. I haven't found a clear description yet...

**Comment by chrisis2015** [comment]
I searched a bit for MySQL catalogs. Is it really relevant to keep it in the join condition at all? 

Have a look at this answer which should explain what a catalog is in relation with schemas and databases in MySQL:

http://stackoverflow.com/questions/7942520/relationship-between-catalog-schema-user-and-database-instance

What do you think?

**Comment by Thorium** [comment]
I think that if we remove the condition then there will be some existing user using a database-backup or a test environment that is a copy of the database restored to different schema, and that will then mess-up that. So this kind of "if-null" could be more near a non-breaking-change.

**Comment by chrisis2015** [comment]
Ok, I understand. When could your suggested fix be implemented? 

By the way, I read this page and the comments on it:

https://bugs.mysql.com/bug.php?id=35427

It seems that the issue I got was there till version 5.5.0 of MySQL.

**Comment by Thorium** [comment]
Fixed in 1.0.47.

**Comment by chrisis2015** [comment]
Thank you Thorium. And I love this library!
